Source/ARTRealtimePresence.m

```diff
@@ -622,4 +660,325 @@ - (void)failPendingPresence:(ARTStatus *)status {
     }
 }
 
+- (void)broadcast:(ARTPresenceMessage *)pm {
+    [_eventEmitter emit:[ARTEvent newWithPresenceAction:pm.action] with:pm];
+}
+
+- (void)sync {
+    [self sync:nil];
+}
+
+- (void)sync:(ARTCallback)callback {
+    if (callback) {
+        ARTCallback userCallback = callback;
+        callback = ^(ARTErrorInfo *__nullable error) {
+            dispatch_async(self->_userQueue, ^{
+                userCallback(error);
+            });
+        };
+    }
+
+    switch (_channel.state) {
+        case ARTRealtimeChannelInitialized:
+        case ARTRealtimeChannelDetaching:
+        case ARTRealtimeChannelDetached: {
+            ARTErrorInfo *error = [ARTErrorInfo createWithCode:ARTErrorBadRequest
+                                                       message:@"Unable to sync to channel; not attached."];
+            ARTLogError(self.logger, @"%@", error.message);
+            if (callback) callback(error);
+            return;
+        }
+        default:
+            break;
+    }
+
+    ARTLogVerbose(self.logger, @"R:%p C:%p (%@) requesting a sync operation", _realtime, _channel, _channel.name);
+
+    ARTProtocolMessage *msg = [[ARTProtocolMessage alloc] init];
+    msg.action = ARTProtocolMessageSync;
+    msg.channel = _channel.name;
+    msg.channelSerial = _syncChannelSerial;
+
+    [self startSync];
+    [_realtime send:msg sentCallback:^(ARTErrorInfo *error) {
+        if (error) {
+            ARTLogDebug(self.logger, @"R:%p C:%p (%@) SYNC request failed with %@", self->_realtime, self->_channel, self->_channel.name, error);
+            [self endSync];
+            if (callback) callback(error);
+        }
+        else {
+            ARTLogDebug(self.logger, @"R:%p C:%p (%@) SYNC requested with success", self->_realtime, self->_channel, self->_channel.name);
+            if (callback) callback(nil);
+        }
+    } ackCallback:nil];
+}
+
+- (void)beforeAttached:(ARTProtocolMessage *)message {
+    if (message.hasPresence) {
+        [self startSync];
+    }
+    else {
+        // RTP1 - when an ATTACHED message is received without a HAS_PRESENCE flag, reset PresenceMap
+        [self startSync];
+        [self endSync];
+        ARTLogDebug(self.logger, @"R:%p C:%p (%@) PresenceMap has been reset", _realtime, self, _channel.name);
+    }
+}
+
+- (void)afterAttached:(ARTProtocolMessage *)message fromState:(ARTRealtimeChannelState)state {
+    if (state == ARTRealtimeChannelAttached) {
```

Comment:
Is this a duplicate code here. We are already handling this before calling `beforeAttached` right?
Also, `if (state == ARTRealtimeChannelAttached) {` will always return `true`.
Can we please simplify this and remove `beforeAttached` and `afterAttached`. I don't think we need two separate methods.
We can keep just one `onAttached` like we do here 
https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L164

Reply:
https://github.com/ably/ably-cocoa/pull/1870/commits/a8d40623b9b6e1d8afba752f91084c0e1f24201f

Reply:
https://github.com/ably/ably-cocoa/pull/1870/commits/bee8c1eb61cf094eb8cd88491c70c713776ca9f8